src/middlewared/middlewared/plugins/smb_sharesec.py

```diff
@@ -0,0 +1,409 @@
+                continue
+
+            acl_entry.update(m.groupdict())
+            if (options.get('resolve_sids', True)) is True:
+                wb = await run([SMBCmd.WBINFO.value, '--sid-to-name', acl_entry['ae_who_sid']], check=False)
+                if wb.returncode == 0:
+                    wb_ret = wb.stdout.decode()[:-3].split('\\')
+                    sidtypeint = int(wb.stdout.decode().strip()[-1:])
+                    acl_entry['ae_who_name'] = {'domain': None, 'name': None, 'sidtype': SIDType.NONE}
+                    acl_entry['ae_who_name']['domain'] = wb_ret[0]
+                    acl_entry['ae_who_name']['name'] = wb_ret[1]
+                    acl_entry['ae_who_name']['sidtype'] = SIDType(sidtypeint).name
+                else:
+                    self.logger.debug(
+                        'Failed to resolve SID (%s) to name: (%s)' % (acl_entry['ae_who_sid'], wb.stderr.decode())
+                    )
+
+            parsed_share_sd['share_acl'].append(acl_entry)
+
+        return parsed_share_sd
+
+    @private
+    async def _sharesec(self, **kwargs):
+        """
+        wrapper for sharesec(1). This manipulates share permissions on SMB file shares.
+        The permissions are stored in share_info.tdb, and apply to the share as a whole.
+        This is in contrast with filesystem permissions, which define the permissions for a file
+        or directory, and in the latter case may also define permissions inheritance rules
+        for newly created files in the directory. The SMB Share ACL only affects access through
+        the SMB protocol.
+        """
+        action = kwargs.get('action')
+        share = kwargs.get('share', '')
+        args = kwargs.get('args', '')
+        sharesec = await run([SMBCmd.SHARESEC.value, share, action, args], check=False)
+        if sharesec.returncode != 0:
+            raise CallError(f'sharesec {action} failed with error: {sharesec.stderr.decode()}')
+        return sharesec.stdout.decode()
+
+    @private
+    async def _delete(self, share):
+        """
+        Delete stored SD for share. This should be performed when share is
+        deleted.
+        If share_info.tdb lacks an entry for the share, sharesec --delete
+        will return -1 and NT_STATUS_NOT_FOUND. In this case, exception should not
+        be raised.
+        """
+        try:
+            await self._sharesec(action='--delete', share=share)
+        except Exception as e:
+            if 'NT_STATUS_NOT_FOUND' not in str(e):
+                raise CallError(e)
+
+    @private
+    async def _view_all(self, options=None):
+        """
+        Return Security Descriptor for all shares.
+        """
+        share_sd_list = []
+        idx = 1
+        share_entries = (await self._sharesec(action='--view-all')).split('\n\n')
+        for share in share_entries:
+            parsed_sd = await self.middleware.call('smb.sharesec._parse_share_sd', share, options)
```

Comment:
This (`smb.sharesec._parse_share_sd`) won't work, middleware does not see these methods